apps/src/templates/codeReviewGroups/CodeReviewGroupsManager.jsx

```diff
@@ -14,57 +14,17 @@ const DROPPABLE_ID_UNASSIGNED = 'unassigned';
 // can be found here:
 // https://github.com/atlassian/react-beautiful-dnd
 export default function CodeReviewGroupsManager({initialGroups}) {
-  const [groups, setGroups] = useState(
+  const [groups, getGroup, setGroups] = useGroups(
     initialGroups.map(group => addDroppableIdToGroup(group))
   );
 
-  const getGroup = droppableId =>
-    _.find(groups, group => group.droppableId === droppableId);
   const getUnassignedGroup = () => getGroup(DROPPABLE_ID_UNASSIGNED);
   const getAssignedGroups = () =>
     groups.filter(group => group.droppableId !== DROPPABLE_ID_UNASSIGNED);
 
-  function onDragEnd(result) {
-    const {source, destination} = result;
-    const sourceId = source.droppableId;
-
-    // dropped outside the group
-    if (!destination) {
-      return;
-    }
-
-    const destinationId = destination.droppableId;
-
-    if (sourceId === destinationId) {
-      // If a member was dropped in its current group.
-      const result = reorder(
-        getGroup(sourceId),
-        source.index,
-        destination.index
-      );
-
-      const updatedGroups = updateGroups(groups, [result]);
-      setGroups(updatedGroups);
-    } else {
-      // If a member was dropped in a different group.
-      const result = move(
-        getGroup(sourceId),
-        getGroup(destinationId),
-        source.index,
-        destination.index
-      );
-
-      const updatedGroups = updateGroups(groups, [
-        result.updatedSource,
-        result.updatedDest
-      ]);
-
-      setGroups(updatedGroups);
-    }
-  }
-
+  // Creating groups no longer works (b/c we want to set groups here...)
   return (
-    <DragDropContext onDragEnd={onDragEnd}>
+    <DragDropContext onDragEnd={setGroups}>
       <div style={styles.dragAndDropContainer}>
         <UnassignedStudentsPanel unassignedGroup={getUnassignedGroup()} />
         <AssignedStudentsPanel
```

Comment:
Can't comment on unchanged lines, but if you look at line 73 that creates new groups, this no longer works because the `setGroups` function we're using is explicitly tied to the structure of the callback expected when ending a drag and drop (ie, dropping something). To create a group, we want to do some setting of state directly (ie, set groups to an array that includes a new element in it). 

What should I do here? Also pass `setGroupsState` function that is defined in the hook to this component as well? I feel like I may just be making things more complicated...

Reply:
i think we need another layer of decoupling to make this work. you've decoupled most of the drag-and-drop logic from this component and put it into the hook, but now the hook is tightly coupled to the one type of `setGroups` that is "setting a group after dragging-and-dropping."

if you look through this file and the hook file, there are a bunch of discrete helper functions -- these should probably be pulled into a utils file, which will make them easily testable and reusable. here are some of the functions i see that should be decoupled and/or moved to their own file: `addDroppableIdToGroup`, `generateNewGroup`, `getGroup`, `updateGroups`, all of the logic currently in `setGroups` that is coupled to the react-beautiful-dnd implementation details, `reorder`, and `move`.

you should probably put these helpers in their own file either way, but that doesn't mean this file has to use those utils; you can expose an API from your hook (rather than using the `[state, setState]` structure). your hook could use the helper functions, and, at the same time, return an object that serves as a more fully-featured API. [this article](https://dev.to/namick/writing-your-own-react-hooks-the-return-value-3lp6) has some good examples for when to return an array or an object from your hook